### PR TITLE
emul: Support UART device emulation

### PR DESCRIPTION
**Pull request by pdgendt**
Add support to the existing UART emulated bus, to have a client device emulator.

### CODE REVIEW

File: drivers/serial/uart_emul.c [hunk 1]
```diff
@@ -19,12 +20,20 @@
...
+
 	bool loopback;
 	size_t latch_buffer_size;
 };
 
+BUILD_ASSERT(offsetof(struct uart_emul_config, emul_list) == 0);
+
 /* Device run time data */
 struct uart_emul_data {
+	/* List of struct spi_emul associated with the device */
```
Comment (cfriedt) [inline drivers/serial/uart_emul.c:34]:
```suggestion
	/* List of struct uart_emul associated with the device */
```

Comment (pdgendt) [reply drivers/serial/uart_emul.c:34]: Oops, copy-pasta

File: include/zephyr/drivers/uart_emul.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+struct uart_emul_device_api;
+
+/**
+ * @brief Define the emulation callback function signature for
+ * uart_emul_device_tx_data_ready_set() function.
+ *
+ * @param dev UART device instance
+ * @param size Number of available bytes in TX buffer
+ * @param user_data Arbitrary user data
```
Comment (cfriedt) [inline include/zephyr/drivers/uart_emul.h:41]: I think changing the `@param` argument should fix the doc build error. Just a suggestion for the description. Feel free to use something better.

```suggestion
 * @param emul pointer to emulation context
```

**Review by cfriedt — COMMENTED** [review]
Looks good - I wonder if I should attempt to convert the gpio emul driver to use this framework somehow.

**Comment by pdgendt** [comment]
> Looks good - I wonder if I should attempt to convert the gpio emul driver to use this framework somehow.

The emul framework, as I see it now, is particularly useful for devices on busses. And UART devices were missing, but not sure how this would work for GPIO.

**Comment by cfriedt** [comment]
> The emul framework, as I see it now, is particularly useful for devices on busses. And UART devices were missing, but not sure how this would work for GPIO.

Technically a wire is a bus (of size 1) and it's possible to attach peripherals to them as well or represent peripheral devices in devicetree.

There is no addressing though :-/

**Comment by pdgendt** [comment]
> Technically a wire is a bus (of size 1) and it's possible to attach peripherals to them as well or represent peripheral devices in devicetree.

But we don\t have in-tree peripherals that are represented as GPIO child nodes, where the PIN index could theoretically be the address. Or do you refer to bitbang drivers? These do implement other bus types, though.

> There is no addressing though :-/

UART also doesn't have addressing, I worked around that in this PR and allow multiple devicetree child nodes even though this will almost never be the case.

**Comment by pdgendt** [comment]
> There are a couple of RF components that are represented, IIRC.
Do you happen to know where?

Another tricky bit that I currently run into is that the emulated UART bus is initialised in PRE_KERNEL level, which is problematic for the emulated init calls if they use kernel objects.

Not sure if a dts property is desired or make it conditional if the UART bus has children or something..

**Comment by pdgendt** [comment]
Ping, now that v3.7 is released, I'd appreciate some reviews.

**Review by cfriedt — APPROVED** [review]
Still LGTM

File: drivers/serial/Kconfig.emul [hunk 3]
```diff
@@ -7,6 +7,8 @@ config UART_EMUL
 	bool "Emulated UART driver [EXPERIMENTAL]"
 	default y
 	depends on DT_HAS_ZEPHYR_UART_EMUL_ENABLED
+	depends on EMUL
```
Comment (dcpleung) [inline drivers/serial/Kconfig.emul:10]: Wonder if `select` would be better here. Then you won't have to add `CONFIG_EMUL=y` in the test config files.

Comment (pdgendt) [reply drivers/serial/Kconfig.emul:10]: I did the same as for the other emulated bus devices.

**Review by dcpleung — APPROVED** [review]

**Pull request merged**